### PR TITLE
feat: improved configuration and sorting of lsp_{document,workspace}_diagnostics

### LINKED ISSUE #846: LSP Diagnostics - Sort by severity
**Is your feature request related to a problem? Please describe.**

Currently LSP diagnostics are mixed between different severity levels (hint, warning, error).

This is quite frustrating since the list might show many unimportant messages first, which makes it hard to focus on the errors.

**Describe the solution you'd like**

Add a custom sorter for diagnostics that first sorts the list by descending severity (error, warning, hint).

### PR DESCRIPTION
**Pull request by fdschmidt93**
Closes #846 

This PR adds various flags to enable explicit selection of diagnostics by severity. This preferred over passing `default_text` as the flag is buggy (see issue) and does not allow to group diagnostics.

For instance, you can now filter for "Error" with

`lua require 'telescope.builtin'.lsp_workspace_diagnostics{severity='error'}` or `lua require 'telescope.builtin'.lsp_workspace_diagnostics{severity=1}` 

and limit the search to `Error` and `Warning` with `lua require 'telescope.builtin'.lsp_workspace_diagnostics{severity_threshold='warning'}`

Further, the PR
* Improves sorting: by bufnr (incl. prioritize current buffer), severity, and line number -- bufnr corresponds to hierarchical file tree of project with highest level files having lowest bufnr (though the initially opened file will have the lowest bufnr, though hunch is this is still preferred over alphabetical file name sorting)
* Link severity definition to `vim.lsp.protocol.DiagnosticSeverity`, just a tiny bit cleaner and more future-proof I suppose
* Completed documentation for `no_sign` and `line_width`

### CODE REVIEW

**Comment by tami5** [comment]
Thanks @fdschmidt93 awesome work!!!!!!, LGTM, let's wait and see what others think and experience < no longer uses Linux 😆.

**Comment by tjdevries** [comment]
Can we change `severity_threshold` -> `severity_limit` to mimic the names from the builin LSP?

**Comment by tjdevries** [comment]
After that, LGTM and we can merge. Thanks!

**Comment by fdschmidt93** [comment]
Thanks @tami5 :)

Done! @tjdevries

File: lua/telescope/utils.lua
Comment (elianiva) [inline lua/telescope/utils.lua:181]: I think it should be indented this way

```suggestion
      return a.bufnr < b.bufnr
    end
  end)
```

File: lua/telescope/builtin/init.lua
Comment (Conni2461) [inline lua/telescope/builtin/init.lua:314]: `string or number` is not a thing in the grammar. Its `string|number|function|table`. Thats probably breaking `docgen` output. (No idea if `docgen` is handling multiple types tho :thinking:)

Comment (fdschmidt93) [reply lua/telescope/builtin/init.lua]: Yeah, does not seem to work. I've not set the type to `string` and added a note for severity fields that they take one of strings or numbers as per `vim.lsp.protocol.DiagnosticSeverity`.

I hope that's fine.

Comment (Conni2461) [reply lua/telescope/builtin/init.lua]: image

Treesitter grammar understands it so only the `docgen` module is missing support. I can add this, that week

Comment (fdschmidt93) [reply lua/telescope/builtin/init.lua]: Ok cool, then let's hold off on this up until then.

Comment (Conni2461) [reply lua/telescope/builtin/init.lua]: https://github.com/tjdevries/tree-sitter-lua/pull/26

File: lua/telescope/builtin/init.lua
Comment (smithbm2316) [inline lua/telescope/builtin/init.lua:314]: What does `wrt` stand for here?

**Review by smithbm2316 — COMMENTED** [review]
Just one small comment in the docs that I'd like clarification, otherwise this LGTM

File: lua/telescope/builtin/init.lua
Comment (fdschmidt93) [reply lua/telescope/builtin/init.lua:314]: `with respect to` - I've abbreviated it to keep the line short enough so linter wouldn't complain :sweat_smile: 

Happy to write it in full if that's preferred.

**Comment by fdschmidt93** [comment]
Docgen now looks clean, thanks again @Conni2461 

If otherwise fine, this can now be merged :) thanks

**Review by Conni2461 — APPROVED** [review]
LGTM Thanks :)

I'll give it a quick try and will then merge

**Pull request merged**
